Add mercy option to score module

Pugzy opened the pull request:
Allow a `mercy` option within the score module. This can be used alongside a `limit` or standalone. As referenced in #250. 

### Example

This XML example will result in a match that progresses like the scoreboard screenshots. 
The winner is the first team to get a mercy lead of 5 or reach the limit of 10.

```xml
<score>
    <mercy>5</mercy>
    <limit>10</limit>
</score>
```

![image](https://user-images.githubusercontent.com/8608892/89710319-cce43700-d979-11ea-85da-c8e65db12480.png)

*Blue team take a storming lead with just one flag away from the win, but wait, red team pulls it back and forces the score limit. 
What an epic gamer moment, so glad I got to watch this historic match on Nartica ™️ .*

### Notes

If a `limit` is not set then a team will need to lead by the set mercy amount or fallback on another condition such as time.

The example above uses CTF but as it is the `ScoreModule` it would work with TDMs and Hills.

This just hooks into the scoring module so no changes to the scoreboard etc. I'd imagine this feature only being used in competitive environments where teams are aware of the score limit and mercy rules (from scrimming), otherwise, the scoreboard limit changing "randomly" may be a bit confusing. Maybe some UI work could be done to make this clearer.

### Code

I have toyed around with the idea of including the feature directly in the score module but decided to separate it out into its own `MercyRule` class so I can have a few helper methods to make it a bit more readable. The topic of micro optimisation was mentioned in the linked issue, the current implementation only re-sorts all team scores if the lead or trail have their score decreased.

Wanted to get the PR made for this for feedback rather than having a useless finished branch that I had not touched in 3 months. 

Signed-off-by: Pugzy <pugzy@mail.com>

Pugzy committed:
Add mercy option to score module

Signed-off-by: Pugzy <pugzy@mail.com>

Pablete1234 commented on core/src/main/java/tc/oc/pgm/score/MercyRule.java:
```diff
@@ -0,0 +1,100 @@
+  private final int scoreLimit;
+  private final int mercyLimit;
+
+  private AbstractMap.SimpleEntry<Competitor, Double> leader;
+  private AbstractMap.SimpleEntry<Competitor, Double> trailer;
```
```suggestion
  private Map.Entry<Competitor, Double> leader;
  private Map.Entry<Competitor, Double> trailer;
```
Just define the generric types here

Pablete1234 reviewed (commented):
One small change, but lgtm

Pugzy commented:
- Updated to allow for negative scoring which will also affect the required score.
- Removed reliance on `CompetitorScoreChangeEvent` and call mercy rules directly from the `incrementScore` method.
- Populate the score module with a default score of 0 for all teams (previously didn't exist in map until points scored).

Electroid reviewed (approved):
👍
